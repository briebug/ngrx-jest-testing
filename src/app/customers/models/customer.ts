export interface Customer {
  id: number;
  name: string;
}
