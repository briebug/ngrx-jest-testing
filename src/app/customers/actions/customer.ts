import { Action } from '@ngrx/store';
import { Customer } from '../models/customer';

export enum CustomerActionTypes {
  Load = '[Customer] Load',
  LoadSuccess = '[Customer] Load Success',
  LoadFail = '[Customer] Load Fail',
  Select = '[Customer] Select'
}

export class Load implements Action {
  readonly type = CustomerActionTypes.Load;
}

export class LoadSuccess implements Action {
  readonly type = CustomerActionTypes.LoadSuccess;

  constructor(public payload: Customer[]) {}
}

export class LoadFail implements Action {
  readonly type = CustomerActionTypes.LoadFail;

  constructor(public error: any) {}
}

export class Select implements Action {
  readonly type = CustomerActionTypes.Select;

  constructor(public payload: Customer) {}
}

export type CustomerActions = Load | LoadSuccess | LoadFail | Select;
