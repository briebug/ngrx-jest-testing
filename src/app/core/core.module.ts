import { CommonModule } from '@angular/common';
import {
  ModuleWithProviders,
  NgModule,
  Optional,
  SkipSelf
} from '@angular/core';

import { CustomersService } from '@core/services/customer.service';

@NgModule({
  imports: [CommonModule],
  providers: [CustomersService]
})
export class CoreModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CoreModule
    };
  }

  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import it in the AppModule only'
      );
    }
  }
}
