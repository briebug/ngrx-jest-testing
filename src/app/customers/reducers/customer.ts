import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from '../models/customer';
import { CustomerActions, CustomerActionTypes } from '../actions/customer';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number | null;
  loading: boolean;
  error: any;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      return { ...state, error: { message: 'Error loading customers' } };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
