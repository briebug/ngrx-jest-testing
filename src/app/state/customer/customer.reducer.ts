import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from './customer.model';
import { CustomerActions, CustomerActionTypes } from './customer.actions';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number;
  loading: boolean;
  error: string;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      console.log(action.error);
      return { ...state, error: 'Error loading customers' };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
