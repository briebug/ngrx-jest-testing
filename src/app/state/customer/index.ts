import { createSelector, createFeatureSelector } from '@ngrx/store';

import * as fromCustomers from './customer.reducer';
import { State as CustomersState } from './customer.reducer';

export const reducers = {
  customers: fromCustomers.reducer
};


export const getCustomersState = createFeatureSelector<CustomersState>(
  'customer'
);

export const {
  selectAll: getAllCustomers,
  selectEntities: getCustomerEntities,
  selectIds: getCustomerIds,
  selectTotal: getTotalCustomers
} = fromCustomers.adapter.getSelectors(getCustomersState);

export const getSelectedCustomerId = createSelector(
  getCustomersState,
  fromCustomers.getSelectedId
);

export const getSelectedCustomer = createSelector(
  getCustomerEntities,
  fromCustomers.getSelectedId,
  (entities, id) => entities[id]
);

export const getLoading = createSelector(
  getCustomersState,
  fromCustomers.getLoading
);

export const getError = createSelector(
  getCustomersState,
  fromCustomers.getError
);
