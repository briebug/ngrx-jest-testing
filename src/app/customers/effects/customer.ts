import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from '../actions/customer';

import { Customer } from '../models/customer';
import { map, switchMap, catchError } from 'rxjs/operators';
import { CustomersService } from '../services/customers';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
