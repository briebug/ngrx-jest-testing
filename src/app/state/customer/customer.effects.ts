import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';
import { map, switchMap, catchError } from 'rxjs/operators';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from './customer.actions';

import { Customer } from './customer.model';
import { CustomersService } from '@core/services/customer.service';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
