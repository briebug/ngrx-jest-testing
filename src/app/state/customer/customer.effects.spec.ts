import { TestBed } from '@angular/core/testing';
import { CustomersService } from '@core/services/customer.service';
import { Actions } from '@ngrx/effects';
import { cold, hot } from 'jasmine-marbles';
import { Observable } from 'rxjs/Observable';
import { empty } from 'rxjs/observable/empty';
import { Load, LoadFail, LoadSuccess } from './customer.actions';
import { CustomerEffects } from './customer.effects';
import { Customer } from './customer.model';

export class TestActions extends Actions {
  constructor() {
    super(empty());
  }

  set stream(source: Observable<any>) {
    this.source = source;
  }
}

export function getActions() {
  return new TestActions();
}

describe('CustomerEffects', () => {
  let effects: CustomerEffects;
  let actions$: TestActions;
  let customersService: any;

  const customer1 = { id: 1, name: 'User 1' } as Customer,
    customer2 = { id: 2, name: 'User 2' } as Customer,
    customers = [customer1, customer2];

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [
        CustomerEffects,
        {
          provide: CustomersService,
          useValue: { load: jest.fn() }
        },
        {
          provide: Actions,
          useFactory: getActions
        }
      ]
    });

    effects = TestBed.get(CustomerEffects);
    customersService = TestBed.get(CustomersService);
    actions$ = TestBed.get(Actions);
  });

  describe('load$', () => {
    it('should return a new customer.LoadSuccess, with the customers, on success', () => {
      const action = new Load(),
        completion = new LoadSuccess(customers),
        response = cold('a|', { a: customers }),
        expected = cold('-b', { b: completion });

      actions$.stream = hot('-a', { a: action });

      // mock the load function to be the response
      customersService.load = jest.fn(() => response);

      expect(effects.load$).toMatchSnapshot();
    });

    it('should return a new customer.LoadError, on fail', () => {
      const action = new Load(),
        completion = new LoadFail('error'),
        response = cold('#'),
        expected = cold('-b', { b: completion });

      actions$.stream = hot('-a', { a: action });

      // mock the load function to be the response
      customersService.load = jest.fn(() => response);

      expect(effects.load$).toMatchSnapshot();
    });
  });
});
