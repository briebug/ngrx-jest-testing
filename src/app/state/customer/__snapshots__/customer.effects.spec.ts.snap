// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerEffects load$ should return a new customer.LoadError, on fail 1`] = `
Actions {
  "_isScalar": false,
  "operator": SwitchMapOperator {
    "project": [Function],
    "resultSelector": undefined,
  },
  "source": Actions {
    "_isScalar": false,
    "operator": FilterOperator {
      "predicate": [Function],
      "thisArg": undefined,
    },
    "source": TestActions {
      "_isScalar": false,
      "source": TestHotObservable {
        "_isScalar": false,
        "error": undefined,
        "marbles": "-a",
        "source": SubscriptionLoggable {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "messages": Array [
            Object {
              "frame": 10,
              "notification": Notification {
                "error": undefined,
                "hasValue": true,
                "kind": "N",
                "value": Load {
                  "type": "[Customer] Load",
                },
              },
            },
          ],
          "observers": Array [],
          "scheduler": TestScheduler {
            "SchedulerAction": [Function],
            "actions": Array [],
            "active": false,
            "assertDeepEqual": [Function],
            "coldObservables": Array [
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 0,
                    "notification": Notification {
                      "error": "error",
                      "hasValue": false,
                      "kind": "E",
                      "value": undefined,
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 10,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": true,
                      "kind": "N",
                      "value": LoadFail {
                        "error": "error",
                        "type": "[Customer] Load Fail",
                      },
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
            ],
            "flushTests": Array [],
            "frame": 0,
            "hotObservables": Array [
              [Circular],
            ],
            "index": -1,
            "maxFrames": 750,
            "now": [Function],
            "scheduled": undefined,
          },
          "subscriptions": Array [],
          "thrownError": null,
        },
        "values": Object {
          "a": Load {
            "type": "[Customer] Load",
          },
        },
      },
    },
  },
}
`;

exports[`CustomerEffects load$ should return a new customer.LoadSuccess, with the customers, on success 1`] = `
Actions {
  "_isScalar": false,
  "operator": SwitchMapOperator {
    "project": [Function],
    "resultSelector": undefined,
  },
  "source": Actions {
    "_isScalar": false,
    "operator": FilterOperator {
      "predicate": [Function],
      "thisArg": undefined,
    },
    "source": TestActions {
      "_isScalar": false,
      "source": TestHotObservable {
        "_isScalar": false,
        "error": undefined,
        "marbles": "-a",
        "source": SubscriptionLoggable {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "messages": Array [
            Object {
              "frame": 10,
              "notification": Notification {
                "error": undefined,
                "hasValue": true,
                "kind": "N",
                "value": Load {
                  "type": "[Customer] Load",
                },
              },
            },
          ],
          "observers": Array [],
          "scheduler": TestScheduler {
            "SchedulerAction": [Function],
            "actions": Array [],
            "active": false,
            "assertDeepEqual": [Function],
            "coldObservables": Array [
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 0,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": true,
                      "kind": "N",
                      "value": Array [
                        Object {
                          "id": 1,
                          "name": "User 1",
                        },
                        Object {
                          "id": 2,
                          "name": "User 2",
                        },
                      ],
                    },
                  },
                  Object {
                    "frame": 10,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": false,
                      "kind": "C",
                      "value": undefined,
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 10,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": true,
                      "kind": "N",
                      "value": LoadSuccess {
                        "payload": Array [
                          Object {
                            "id": 1,
                            "name": "User 1",
                          },
                          Object {
                            "id": 2,
                            "name": "User 2",
                          },
                        ],
                        "type": "[Customer] Load Success",
                      },
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
            ],
            "flushTests": Array [],
            "frame": 0,
            "hotObservables": Array [
              [Circular],
            ],
            "index": -1,
            "maxFrames": 750,
            "now": [Function],
            "scheduled": undefined,
          },
          "subscriptions": Array [],
          "thrownError": null,
        },
        "values": Object {
          "a": Load {
            "type": "[Customer] Load",
          },
        },
      },
    },
  },
}
`;
